### PR TITLE
Prevent updates/reboots with locks when HUP breadcrumbs present

### PR DESCRIPTION
**Pull request by cywang117**
On HUP rollback, some services need to complete before
it's safe for the Supervisor to reboot the device when
applying state changes. rollback-{health|altboot}-breadcrumb
are the two files that Supervisor locks the device on when
present in this patch.

Not closing issue 1459 because there is a possible case where,
on altboot rollback, the breadcrumbs are not present. 1459
may be closed when this edge case is investigated.

Added some update-lock.ts tests.

Needs some more integration testing, but adding the rollback files manually in 
`/mnt/state` causes the update locks to behave as expected. Will integration test
more tomorrow.

Change-type: patch
Connects-to: #1459
Signed-off-by: Christina Wang <christina@balena.io>

### CODE REVIEW

File: test/src/lib/update-lock.spec.ts [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+		beforeEach(() => {
+			// @ts-ignore
+			unlockSyncStub = stub(lockFile, 'unlockSync').callsFake((lockPath) => {
+				// Throw error on process.exit for one of the two lockpaths
+				if (lockPath === testLockPaths[1]) {
+					throw new Error(
+						'handled unlockSync error which should not crash test process',
+					);
+				}
+			});
```
Comment (cywang117) [inline test/src/lib/update-lock.spec.ts:115]: This tests both error thrown + no error thrown case when attempting to cleanup lockfiles on Node process exit. The error thrown case should be handled silently by the code currently.

File: test/src/lib/update-lock.spec.ts [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+				constants.rootMountPoint,
+				updateLock.lockPath(appId),
+				service,
+			)]: {
+				'updates.lock': '',
+			},
+		});
+	};
+
+	// TODO: Remove these hooks when we don't need './test/data' as test process's rootMountPoint
```
Comment (cywang117) [inline test/src/lib/update-lock.spec.ts:42]: These `before`/`after` hooks can be removed when every test that interacts with the fs in some way is switched over to mock-fs instead of using the physical file system.

**Comment by 20k-ultra** [comment]
@cywang117 can you ping for review again once the integration tests are done. I'm worried if significant changes will be made then we have to review again.

**Comment by 20k-ultra** [comment]
How is the integration testing going ?

**Comment by cywang117** [comment]
Commented in Flowdock, I'll paste my comment here for reference:

Quick update regarding the HUP breadcrumbs reboot lock PR. I've figured out why I've been having trouble testing it on the TX2 now! I'd previously added breadcrumb checks in `updateLock.lock` which is fine, and this function gets called by `applicationManager.lockingIfNecessary`, but when you follow the code flow upwards to `deviceState.reboot`, the lock function is only reached when there are services and an application present on the device, which the TX2 did not have. This is why I couldn't see the behavior no matter what I tried. It's pretty obvious in hindsight and I'm shaking my head at myself slightly (haha 😢), but now I know how to modify the code to separate the HUP breadcrumb checking functionality from the service-level lockfile checking functionality. 

I'll be making updates to the PR early next week and integration testing it locally then on Ross's TX2, and I should have a working PR ready for review by late next week provided there are no hiccups. (Fingers crossed!)

As far as unit tests go, I think I just need to add a case to test for prevention of deviceState.reboot or deviceState.shutdown when breadcrumbs are present. All else is actual device testing (I've written up a how-to and a convenience script for testing this and I'll share those when I'm done with everything)

**Comment by cywang117** [comment]
@20k-ultra @pipex This PR is ready for review! Integration tests completed successfully. [Here](https://www.notion.so/HUP-rollback-tests-6827d84a7f6b4d798ac0ceccb88582a4) is the process I followed.

image

File: src/device-state.ts [hunk 3]
```diff
@@ -647,6 +648,7 @@ export async function reboot(force?: boolean, skipLock?: boolean) {
 }
 
 export async function shutdown(force?: boolean, skipLock?: boolean) {
+	await updateLock.ensureNoHUPBreadcrumbsOnHost();
```
Comment (cywang117) [inline src/device-state.ts:651]: Adding the HUP checks here instead of in updateLock.lock. That function is currently only entered when there is an application running on the device, because `lockingIfNecessary` is part of `applicationManager`. However when no app / service is present, updateLock.lock never gets called, hence why I was having trouble with integration tests before.

File: test/05-device-state.spec.ts [hunk 4]
```diff
@@ -355,4 +356,26 @@ describe('deviceState', () => {
...
+			.then(() => expect.fail())
+			.catch((err: any) => {
+				expect(err.message).to.equal(testErrMsg);
+			});
+		await deviceState
+			.shutdown()
+			.then(() => expect.fail())
+			.catch((err: any) => {
+				expect(err.message).to.equal(testErrMsg);
+			});
```
Comment (20k-ultra) [inline test/05-device-state.spec.ts:377]: I think this will fail if the promise does not throw but also passes when it does throw and matches our error (also matches on the error type which is nice since some logic does err instanceof)

`
await expect(deviceState.reboot()).to.eventually.throw(new UpdatesLockedError(testErrMsg));
await expect(deviceState.shutdown()).to.eventually.throw(new UpdatesLockedError(testErrMsg));
`

Comment (20k-ultra) [reply test/05-device-state.spec.ts:377]: feel free to see if this cleans up any other test case code such as the update-lock spec

File: test/src/lib/update-lock.spec.ts [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+import * as Bluebird from 'bluebird';
+
+import rewire = require('rewire');
+import mockFs = require('mock-fs');
+
+import * as constants from '../../../src/lib/constants';
+import { UpdatesLockedError } from '../../../src/lib/errors';
+
+describe('lib/update-lock', () => {
+	const updateLock = rewire('../../../src/lib/update-lock');
```
Comment (20k-ultra) [inline test/src/lib/update-lock.spec.ts:13]: glad to see rewire being useful for others as well. I've really enjoyed it.

File: src/lib/update-lock.ts [hunk 6]
```diff
@@ -35,6 +36,29 @@ function lockFilesOnHost(appId: number, serviceName: string): string[] {
...
+ * rollback-{health|altboot}.service have not exited. If these files exist,
+ * prevent reboot. If the Supervisor reboots while those services are still running,
+ * the device may become stuck in an invalid state during HUP.
+ */
+export function ensureNoHUPBreadcrumbsOnHost(): Bluebird<boolean | never> {
+	return Bluebird.all(
+		[
+			'rollback-health-breadcrumb',
+			'rollback-altboot-breadcrumb',
+		].map((filename) =>
```
Comment (20k-ultra) [inline src/lib/update-lock.ts:50]: can we replace this bluebird call is [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) ?

Comment (20k-ultra) [reply src/lib/update-lock.ts:50]: I think you did it since the file had bluebird all over it anyways. I don't want to push to scope of this PR but we should refactor this file to remove bluebird. Especially since the lock function is callback hell.

**Review by 20k-ultra — COMMENTED** [review]
killed it! Do you know what the Supevisor API response is when you try to hit shutdown or reboot from the dashboard while these breadcrumbs exist ? Check that out then you can self certify.

I'd say let's hold off on the refactor for another PR to prevent scope creep here. This is already a substantial change so we don't want to introduce too much at once. Feel free to refactor bluebird out and remove the callback hell from lock() if you want after this.

**Comment by 20k-ultra** [comment]
@balena-ci rebase

File: src/lib/update-lock.ts (hunk 6 above)
Comment (cywang117) [reply src/lib/update-lock.ts:50]: I totally forgot to switch out Bluebird for Promise here. Explanation, this function was just used inside updateLock.lock before, which uses Bluebird promises because of `.catchReturn` and `.disposer` which native Promises don't implement. I've since removed ensureNoHUP... from updateLock so Bluebird promises aren';t needed anymore. Thanks for the reminder!

File: test/src/lib/update-lock.spec.ts (hunk 5 above)
Comment (cywang117) [reply test/src/lib/update-lock.spec.ts:13]: Same! It's been invaluable in tests.

File: test/05-device-state.spec.ts (hunk 4 above)
Comment (cywang117) [reply test/05-device-state.spec.ts:377]: `await expect(deviceState.shutdown()).to.eventually.throw(new UpdatesLockedError(testErrMsg));` doesn't work but removing the async/await pattern from the test case does:
```
expect(deviceState.reboot()).to.eventually.throw(new UpdatesLockedError(testErrMsg));
expect(deviceState.shutdown()).to.eventually.throw(new UpdatesLockedError(testErrMsg));
```

Verified that the stub restore call after these 2 expects is called as well. Chai-as-promised handles the underlying promise so no need for async/await I guess.

Comment (cywang117) [reply test/05-device-state.spec.ts:377]: That's weird, because some other test cases use async/await no problem for a setup similar to this..

Comment (cywang117) [reply test/05-device-state.spec.ts:377]:
```
await expect(deviceState.reboot()).to.be.rejectedWith(testErrMsg);
await expect(deviceState.shutdown()).to.be.rejectedWith(testErrMsg);
```
works. Gotta remember to use `rejectedWith` instead of `throw` (note that `rejectedWith(new UpdatesLockedError(testErrMsg)` will not pass). I'm almost certain I've dealt with this exact scenario before with chai-as-promised while writing tests, which means I almost prefer the non chai-as-promised method I had originally:
```
await promise()
.then(() => expect.fail())
.catch(err => {
  expect(err).to.be.instanceOf(UpdatesLockedError);
  expect(err.message).to.equal(testErrMsg);
})
```

Comment (cywang117) [reply test/05-device-state.spec.ts:377]: Already committed to using chai-as-promised though because I spent time on it, so I'll change it 👍🏼

**Comment by cywang117** [comment]
@20k-ultra I removed Bluebird from updateLock.ensureNoHUPBreadcrumbsExist but not the updateLock.lock function, since we still need Bluebird there because of the .disposer util which native Promises doesn't have. I'd love to remove this promise hell (and looked into it while coding this PR) but we'd need a good native alternative.

**Comment by cywang117** [comment]
> Do you know what the Supevisor API response is when you try to hit shutdown or reboot from the dashboard while these breadcrumbs exist?
```
root@92f8b89:/mnt/state# curl -X POST --header "Content-Type:application/json" "10.114.104.1:48484/v1/reboot?apikey=dc28
27d153e94e92a9d2f5ab7a2cda5b" --data '{"force": true}'
{"Data":"","Error":"Waiting for Host OS update to finish"}
```
Looks good to me :)

**Comment by cywang117** [comment]
@balena-ci I self-certify!
